### PR TITLE
Refactor parts of FilesController to use a PosixFile model

### PR DESCRIPTION
**Pull request by robinkar**
Related to #1187.

This PR moves posix file only methods out from the Files model into a PosixFile model instead, which was most of the methods.
The PosixFile model uses instance methods in most places instead of the class methods as it makes handling other files than posix files easier in the future and avoids needing to pass the path to every method.
One could for example parse the request or path to instantiate something else than the PosixFile model, which would then implement the same methods. As a small experiment on this, I implemented a prototype that uses Rclone as a backend for listing files(e.g. #1188), which was surprisingly easy to do.

There are more changes that could be done (e.g. zip handling code and the rest of the FileController methods), but I tried to keep the PR somewhat small to get some feedback first whether this is a reasonable approach.

Suggestions on improvements or further additions are very much welcome.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201735133575781/1201793941169017) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by johrstrom** [comment]
Hey thanks for the PR. It's going to take a bit for me to look this over.

At a glance it looks OK, though I'm wondering about the Files class still being there and PosixFile not inheriting it.  I get that maybe it's too soon for any abstraction.  I'll have to check who's using Files and for what to see what sort of delineation there is between these too.

File: apps/dashboard/app/controllers/files_controller.rb [hunk 1]
```diff
@@ -6,11 +6,13 @@ def fs
     request.format = 'json' if request.headers['HTTP_ACCEPT'].split(',').include?('application/json')
 
     @path = normalized_path
+
     AllowlistPolicy.default.validate!(@path)
 
+    path = PosixFile.new(@path)
```
Comment (johrstrom) [inline apps/dashboard/app/controllers/files_controller.rb:12]: I don't know if I like having both the member variable `@path` and the method scoped variable `path`. As you can see the member variable `@path` is being used in the views and more in this method.

Can we make it so `@path = PosixFile.new(normalized_path)` and `delegate` however many of those methods to _it's_ member variable (like `@path.descend` used in the view).

File: apps/dashboard/app/models/posix_file.rb [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+  def self.num_files(from, names)
+    args = names.map {|n| Shellwords.escape(n) }.join(' ')
+    o, e, s = Open3.capture3("find 2>/dev/null #{args} | wc -l", chdir: from)
+
+    # FIXME: handle status error
+    o.lines.last.to_i
+  end
+
+  def self.username(uid)
```
Comment (johrstrom) [inline apps/dashboard/app/models/posix_file.rb:134]: Can you move all these class methods to a block at the top of the file? I know you probably copied as is - but it seems like we may as well take the opportunity to clean it up a little bit. 

That way we can group all the class methods together and it's a little more organized.

**Comment by robinkar** [comment]
Thanks. Just fixed the things you mentioned.

File: apps/dashboard/app/models/posix_file.rb [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    path.parent
+  end
+
+  def join(*args)
+    path.join(*args)
+  end
+
+  def to_s
+    path.to_s
+  end
```
Comment (johrstrom) [inline apps/dashboard/app/models/posix_file.rb:170]: You can just use Ruby's delegate directly. I'm suggesting this change but can you put it at the top near `attr_reader path`.

```suggestion
delegate :basename, :descend, :parent, :join, :to_s to: :path
```

Comment (robinkar) [reply apps/dashboard/app/models/posix_file.rb]: Fixed.

**Review by johrstrom — APPROVED** [review]
Thanks!

**Pull request merged**